### PR TITLE
Feature/itly 1041 improved validation support v2

### PR DESCRIPTION
**Pull request by ondrej**

### CODE REVIEW

File: packages/plugin-iteratively/src/main/kotlin/ly/iterative/itly/iteratively/IterativelyPlugin.kt [hunk 1]
```diff
@@ -114,37 +114,28 @@ class IterativelyPlugin(
         )
     }
 
-    override fun group(userId: String?, groupId: String, properties: Properties?) {
+    override fun postGroup(userId: String?, groupId: String, properties: Properties?, validationResults: List<ValidationResponse>) {
         this.push(this.toTrackModel(
             type = TrackType.group,
             properties = properties,
-            validation = null
+            validation = validationResults.find { !it.valid }
```
Comment (ondrej) [inline packages/plugin-iteratively/src/main/kotlin/ly/iterative/itly/iteratively/IterativelyPlugin.kt:121]: We could also combine all the messages into one long string to send to Stream.

File: packages/plugin-schema-validator/src/test/kotlin/ly/iterative/itly/SchemaValidatorPluginTest.kt [hunk 2]
```diff
@@ -69,36 +69,53 @@ class SchemaValidatorPluginTest {
...
-            validation = OPTIONS_ERROR_ON_INVALID
+            validation = VALIDATION_OPTIONS_ERROR_ON_INVALID
         ))
 
         val exception = Assertions.assertThrows(IllegalArgumentException::class.java) {
-            itly.validate(invalidEvent)
+            itly.track(null, invalidEvent)
         }
 
+        println("HERE!!!!!")
```
Comment (ondrej) [inline packages/plugin-schema-validator/src/test/kotlin/ly/iterative/itly/SchemaValidatorPluginTest.kt:79]: Lol, tried "console.log debugging" before figuring out how to debug the tests in Android Studio. Will remove on next check-in.

File: packages/sdk-core/src/main/kotlin/ly.iterative.itly/PluginBase.kt [hunk 3]
```diff
@@ -9,18 +9,24 @@ abstract class PluginBase(
         return id
     }
     override fun load(options: Options) {}
+
+    override fun validate(event: Event): ValidationResponse? {
+        return null
```
Comment (ondrej) [inline packages/sdk-core/src/main/kotlin/ly.iterative.itly/PluginBase.kt:14]: Not sure about this but agree it sucks to have to create new objects all the time.

File: packages/sdk-core/src/main/kotlin/ly.iterative.itly/core/Itly.kt [hunk 4]
```diff
@@ -174,128 +100,109 @@ open class Itly {
...
+        postMethod: (plugin: Plugin, event: Event, validationResponses: List<ValidationResponse>) -> Unit
+    ) {
+        val contextValidationResponses = if (includeContext) validate(context) else listOf()
+        val isContextValid = contextValidationResponses.all { it.valid }
+
+        val eventValidationResponses = validate(event)
+        val isEventValid = eventValidationResponses.all { it.valid }
+
+        val combinedEvent = if (includeContext)
+            Event(
```
Comment (ondrej) [inline packages/sdk-core/src/main/kotlin/ly.iterative.itly/core/Itly.kt:181]: Is this the best way to combine the props? Also sucks we have to create a brand new event but couldn't think of anything better.

File: packages/plugin-schema-validator/src/test/kotlin/ly/iterative/itly/SchemaValidatorPluginTest.kt [hunk 5]
```diff
@@ -69,36 +69,53 @@ class SchemaValidatorPluginTest {
...
+        System.err.println("Hello standard error!")
+        println(exception.message)
+
         Assertions.assertEquals(invalidEventExpectedErrorMessage, exception.message)
     }
 
     @Test
-    fun itlyLoad_invalidContextWithErrorOnInvalid_throwsError() {
-        val exception = Assertions.assertThrows(IllegalArgumentException::class.java) {
+    fun itlyLoad_invalidContext_succeeds() {
```
Comment (justin-iteratively) [inline packages/plugin-schema-validator/src/test/kotlin/ly/iterative/itly/SchemaValidatorPluginTest.kt:87]: Should throw on invalid right?

File: packages/sdk-core/src/main/kotlin/ly.iterative.itly/PluginBase.kt [hunk 6]
```diff
@@ -9,18 +9,24 @@ abstract class PluginBase(
         return id
     }
     override fun load(options: Options) {}
+
+    override fun validate(event: Event): ValidationResponse {
+        return ValidationResponse.Valid
```
Comment (justin-iteratively) [inline packages/sdk-core/src/main/kotlin/ly.iterative.itly/PluginBase.kt:14]: 👍

File: packages/sdk-core/src/main/kotlin/ly.iterative.itly/core/Itly.kt (hunk 4 above)
Comment (justin-iteratively) [reply packages/sdk-core/src/main/kotlin/ly.iterative.itly/core/Itly.kt:176]: Maybe `event.apply { properties = properties + context.properties }`?

File: packages/plugin-schema-validator/src/test/kotlin/ly/iterative/itly/SchemaValidatorPluginTest.kt (hunk 5 above)
Comment (justin-iteratively) [reply packages/plugin-schema-validator/src/test/kotlin/ly/iterative/itly/SchemaValidatorPluginTest.kt:87]: Nevermind, I see context is now validated with track

File: packages/plugin-schema-validator/src/test/kotlin/ly/iterative/itly/SchemaValidatorPluginTest.kt (hunk 2 above)
Comment (justin-iteratively) [reply packages/plugin-schema-validator/src/test/kotlin/ly/iterative/itly/SchemaValidatorPluginTest.kt]: Removed

**Review by justin-iteratively — APPROVED** [review]

**Pull request merged**